### PR TITLE
Ensemble utility for optimizing parameters

### PR DESCRIPTION
**Pull request by hunse**
I added a utility for doing a simple random search of ensemble parameters to best decode a given function.

### CODE REVIEW

**Comment by drasmuss** [comment]
This makes me think, it should be possible to write some fairly easy wrappers to convert our distributions into the distributions hyperopt uses (just a matter of mapping our uniform to their uniform, etc.).  Then you could use basically this same function, but with the fancy search routines of hyperopt rather than random search.  Not suggesting that for this PR, just noting it down while I think of it.

**Comment by hunse** [comment]
Yeah, that's a good idea, and should be pretty straightforward (though UniformHypersphere might be a bit tricky, maybe?). I'd still want to have this basic random one, in case people don't have hyperopt installed or don't want to get into it, but it would be great to have that option. I was also wondering whether any more traditional optimization techniques would be applicable (e.g. stochastic gradient descent), but probably not without applying backprop to the whole encoders->neurons->decoders chain.

**Comment by Seanny123** [comment]
LGTM.

**Comment by drasmuss** [comment]
I noticed we were redoing that `sample` function a few places in the code, so I factored it out into `utils`.  I made it a separate commit, because it isn't directly related to this PR, just inspired by it.

Only significant difference is that I used `np.array` instead of `np.asarray`.  We could add a `copy` flag to set which behaviour we want if someone thinks that's important, but as a default I figured I'd just go for the safer option.

**Comment by drasmuss** [comment]
There might be an argument for putting this hyperparameter optimization in `nengo_extras` instead.  No strong feelings either way.

**Comment by Seanny123** [comment]
FYI @drasmuss this is failing flake8

**Comment by jgosmann** [comment]
Might this be better placed in `nengo_extras`?

**Comment by drasmuss** [comment]
That's 2.5 votes for moving this to `nengo_extras` instead, any objections @hunse?

**Comment by hunse** [comment]
Yeah, sounds good to me.

**Comment by jgosmann** [comment]
This mostly LGTM. I think we should separate @drasmuss's factoring out of the `sample` function as this has to be done on the Nengo side. The rest should go into `nengo_extras`.

Only thing that makes me a little bit worried is that this depends on a number of functions from the builder code and I'm not sure how public that API is supposed to be.

Would it make sense to cache decoders in this function? Or isn't it worth the added complexity? (Probably one should save the result of this function somewhere instead of doing this optimization in every build.)

**Comment by tbekolay** [comment]
I've now split this into #1181 and https://github.com/nengo/nengo_extras/pull/26. Closing!
